/*
 * Diplom ‚Äî —à–∏—Ñ—Ä–æ–≤–∞–Ω–∏–µ —Ñ–∞–π–ª–æ–≤ –ø–æ –ì–û–°–¢ —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º –ö—É–∑–Ω–µ—á–∏–∫–∞ –∏ –ú–∞–≥–º—ã
 * Copyright (C) 2025 –û–ª–µ–≥ –£—Å–æ–ª—å—Ü–µ–≤ <jeep2036@mail.ru>
 *
 * –≠—Ç–æ—Ç –ø—Ä–æ–≥—Ä–∞–º–º–Ω—ã–π –æ–±–µ—Å–ø–µ—á–µ–Ω–∏–µ–º —Ä–∞—Å–ø—Ä–æ—Å—Ç—Ä–∞–Ω—è–µ—Ç—Å—è –Ω–∞ —É—Å–ª–æ–≤–∏—è—Ö
 * GNU General Public License –≤–µ—Ä—Å–∏–∏ 3 –∏–ª–∏ –±–æ–ª–µ–µ –ø–æ–∑–¥–Ω–µ–π.
 * –ü–æ–¥—Ä–æ–±–Ω–µ–µ: https://www.gnu.org/licenses/gpl-3.0
 */
#include "passworddialog.h"
#include "crypto/striborg.h"
#include <QRandomGenerator>
#include <QMessageBox>
#include <QRegularExpression>

PasswordDialog::PasswordDialog(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::PasswordDialog),
    accepted(false)
{
    ui->setupUi(this);
    setWindowTitle("–ü–∞—Ä–æ–ª—å");

    ui->passwordEdit->setEchoMode(QLineEdit::Password);
    ui->spinBox->setMinimum(8);
    ui->spinBox->setMaximum(128);
    ui->spinBox->setValue(16);

    ui->lowercaseCheckBox->setChecked(true);
    ui->uppercaseCheckBox->setChecked(true);
    ui->digitsCheckBox->setChecked(true);
    ui->specialCheckBox->setChecked(true);

    ui->buttonBox->button(QDialogButtonBox::Ok)->setText("–ü—Ä–∏–Ω—è—Ç—å");
    ui->buttonBox->button(QDialogButtonBox::Cancel)->setText("–û—Ç–º–µ–Ω–∞");

    connect(ui->passwordEdit, &QLineEdit::textChanged,
            this, &PasswordDialog::on_passwordEdit_textChanged);

    updateStrengthIndicator(0);
}


PasswordDialog::~PasswordDialog()
{
    delete ui;
}

QString PasswordDialog::getPassword() const
{
    return ui->passwordEdit->text().trimmed();
}

bool PasswordDialog::wasAccepted() const
{
    return accepted;
}

QString PasswordDialog::generatePassword()
{
    int length = ui->spinBox->value();
    QString chars;

    if (ui->lowercaseCheckBox->isChecked())
        chars += "abcdefghijklmnopqrstuvwxyz";
    if (ui->uppercaseCheckBox->isChecked())
        chars += "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    if (ui->digitsCheckBox->isChecked())
        chars += "0123456789";
    if (ui->specialCheckBox->isChecked())
        chars += "!#$%&()*+-.:/;<=>?@[]^_`{|}~";

    if (chars.isEmpty())
        return "";

    QString pass;
    for (int i = 0; i < length; ++i) {
        int index = QRandomGenerator::global()->bounded(chars.size());
        pass.append(chars[index]);
    }
    return pass;
}

void PasswordDialog::on_generateButton_clicked()
{
    QString pass = generatePassword();
    if (pass.isEmpty()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í—ã–±–µ—Ä–∏—Ç–µ —Ö–æ—Ç—è –±—ã –æ–¥–∏–Ω –Ω–∞–±–æ—Ä —Å–∏–º–≤–æ–ª–æ–≤.");
        return;
    }
    ui->passwordEdit->setText(pass);
    ui->passwordEdit->setFocus();
}


void PasswordDialog::on_toggleVisible_toggled(bool checked)
{
    ui->passwordEdit->setEchoMode(checked ? QLineEdit::Normal : QLineEdit::Password);
}


int PasswordDialog::checkPasswordStrength(const QString &pass)
{
    if (pass.length() < 8) return 0;

    int score = 0;
    if (pass.length() >= 12) score++;
    if (pass.contains(QRegularExpression("[a-z]"))) score++;
    if (pass.contains(QRegularExpression("[A-Z]"))) score++;
    if (pass.contains(QRegularExpression("[0-9]"))) score++;
    if (pass.contains(QRegularExpression("[#$%&()*+\\-.:/;<=>?@\\[\\]^_`{|}~]"))) score++;

    return score;
}



void PasswordDialog::updateStrengthIndicator(int strength)
{
    QLabel *label = ui->strengthLabel_2;  // ‚Üê –∏—Å–ø—Ä–∞–≤–ª–µ–Ω–æ –Ω–∞ strengthLabel_2
    if (!label) return;

    if (strength < 3) {
        label->setText("–°–ª–∞–±—ã–π");
        label->setStyleSheet("color: red;");
    } else if (strength < 5) {
        label->setText("–°—Ä–µ–¥–Ω–∏–π");
        label->setStyleSheet("color: orange;");
    } else {
        label->setText("–ù–∞–¥—ë–∂–Ω—ã–π");
        label->setStyleSheet("color: green;");
    }
}



void PasswordDialog::on_passwordEdit_textChanged(const QString &text)
{
    int strength = checkPasswordStrength(text);
    updateStrengthIndicator(strength);
}
void PasswordDialog::on_buttonBox_accepted()
{
    QString pass = ui->passwordEdit->text().trimmed();
    if (pass.isEmpty()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ü–∞—Ä–æ–ª—å –Ω–µ –º–æ–∂–µ—Ç –±—ã—Ç—å –ø—É—Å—Ç—ã–º.");
        return;
    }

    int strength = checkPasswordStrength(pass);
    if (strength < 3) {
        int ret = QMessageBox::warning(this, "–°–ª–∞–±—ã–π –ø–∞—Ä–æ–ª—å",
                                       "–ü–∞—Ä–æ–ª—å —Å–ª–∏—à–∫–æ–º —Å–ª–∞–±—ã–π. –í—Å—ë —Ä–∞–≤–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å?",
                                       QMessageBox::Yes | QMessageBox::No);
        if (ret == QMessageBox::No)
            return;
    }

    // üîê –•—ç—à–∏—Ä—É–µ–º –ø–∞—Ä–æ–ª—å —á–µ—Ä–µ–∑ Streebog (256 –∏–ª–∏ 512)
    Streebog streebog(256);  // –∏–ª–∏ 512 ‚Äî –∫–∞–∫ —Ç—Ä–µ–±—É–µ—Ç—Å—è
    QByteArray passUtf8 = pass.toUtf8();
    unsigned char* hash = streebog.hash((unsigned char*)passUtf8.data(), passUtf8.size());

    // –ö–æ–ø–∏—Ä—É–µ–º —Ö—ç—à –≤ QByteArray (32 –±–∞–π—Ç–∞ –¥–ª—è Streebog-256)
    QByteArray hashBytes;
    for (int i = 0; i < 32; ++i) {
        hashBytes.append(static_cast<char>(hash[i]));
    }

    // –°–æ—Ö—Ä–∞–Ω—è–µ–º —Ö—ç—à (–Ω–∞–ø—Ä–∏–º–µ—Ä, –≤ –ø–æ–ª–µ –∫–ª–∞—Å—Å–∞)
    m_passwordHash = hashBytes;

    accepted = true;
    accept();
}
QByteArray PasswordDialog::getPasswordHash() const
{
    return m_passwordHash;
}


void PasswordDialog::on_buttonBox_rejected()
{
    accepted = false;
    reject();
}
